rfcs/0003-template-compiler.md

```diff
@@ -0,0 +1,173 @@
+- Provide a rollup plugin to compile templates.
+- Compiler can be run on JavaScript and TypeScript files.
+- Compilation preserves source-maps.
+
+### Non-Goals
+
+- The compiled template does not have to be syntactically identical to templates prepared at runtime. For example comment markers will not be the same as runtime preparation.
+- Other template transformations. This is only for performance.
+
+## Motivation
+
+`lit-html` has [three internal rendering phases](https://github.com/lit/lit/blob/main/dev-docs/design/how-lit-html-works.md#summary-of-lit-html-rendering-phases) which are all optimized to cache and reuse previous work. However, the very first `render()` must still execute all three phases.
+
+The @lit-labs/compiler will provide a way to opt into precompiling the templates, moving the first internal rendering phase to build time.
+
+By precompiling `html` tagged template literals, the runtime only needs to execute two internal rendering phases reducing the first render performance. As a nice benefit, any subsequent updates that encounter a new compiled `html` tagged template literal will also benefit from the performance improvement and can skip the prepare phase.
+
+## Detailed Design
+
+At a high level, the `@lit-labs/compiler` package provides a transformer that converts:
+
+```js
+const hi = (name) => html`<h1>Hello ${name}!</h1>`;
+```
+
+into:
+
+```js
+const compiled_html = (s) => s;
+const lit_template_1 = {
+  h: compiled_html`<h1>Hello <?></h1>`,
+  parts: [{ type: 2, index: 1 }],
+};
+const hi = (name) => ({ _$litType$: lit_template_1, values: [name] });
+```
+
+This transform converts the `` html`<h1>Hello ${name}!</h1>` `` `TemplateResult` into a `CompiledTemplateResult` which semantically behaves the same but with greater initial render performance. The update performance is unchanged.
+
+### Transformer Design
+
+The transformer should be completely syntactic. Initially only compiling any `html` template expressions if `html` was imported from `lit` or `lit-html` directly.
+
+Each individual `TemplateResult` is split into two declarations, a `CompiledTemplate` and `CompiledTemplateResult`. The `CompiledTemplate` is hoisted to the module scope, and contains the prepared HTML and template parts for the original `TemplateResult`. The `CompiledTemplateResult` replaces the `TemplateResult` and binds the static `CompiledTemplate` with the dynamic `values` array.
+
+If a part is used that requires a constructor, the compiler
+will also insert an import such as:
+
+```js
+// ...
+import { _$LH as litHtmlPrivate_1 } from "lit-html/private-ssr-support.js";
+const {
+  AttributePart: _$LH_AttributePart,
+  PropertyPart: _$LH_PropertyPart,
+  BooleanAttributePart: _$LH_BooleanAttributePart,
+  EventPart: _$LH_EventPart,
+} = litHtmlPrivate_1;
+```
+
+This transformer will be exported from `@lit-labs/compiler`, and can be used anywhere TypeScript transformers are accepted, which depends on build setup.
+
+### Rollup plugin
```

Comment:
Show usage of the TypeScript transformer directly?

Reply:
Done